patches/server/0925-Paper-Plugins.patch

```diff
@@ -0,0 +1,1274 @@
++        }
++
++        return builder.build();
++    }
++
++    private static TextColor fromStatus(PluginProvider<?> provider) {
++        if (provider instanceof ProviderStatusHolder statusHolder && statusHolder.getLastProvidedStatus() != null) {
++            ProviderStatus status = statusHolder.getLastProvidedStatus();
++
++            // Handle enabled/disabled game plugins
++            if (status == ProviderStatus.INITIALIZED && provider instanceof JavaPluginProvider javaPluginProvider) {
++                Plugin plugin = Bukkit.getPluginManager().getPlugin(javaPluginProvider.getConfiguration().getIdentifier());
++                // Plugin doesn't exist? Could be due to it being removed.
++                if (plugin == null) {
++                    return NamedTextColor.RED;
++                }
++
++                return plugin.isEnabled() ? NamedTextColor.GREEN : NamedTextColor.RED;
++            }
++
++            return switch (status) {
++                case INITIALIZED -> NamedTextColor.GREEN;
++                case ERRORED -> NamedTextColor.RED;
++            };
++        } else if (provider instanceof PaperPluginProvider.PaperServerPluginProvider serverPluginProvider && serverPluginProvider.shouldSkipCreation()) {
++            // Paper plugins will be skipped if their provider is skipped due to their initializer failing.
++            // Show them as red
++            return NamedTextColor.RED;
++        } else {
++            return NamedTextColor.GRAY;
++        }
++    }
++
++}
+diff --git a/src/main/java/io/papermc/paper/plugin/provider/PluginProviders.java b/src/main/java/io/papermc/paper/plugin/provider/PluginProviders.java
+new file mode 100644
+index 0000000000000000000000000000000000000000..277e8b3802d5cb07893be91a27f31b8a349fc691
+--- /dev/null
++++ b/src/main/java/io/papermc/paper/plugin/provider/PluginProviders.java
+@@ -0,0 +1,67 @@
++package io.papermc.paper.plugin.provider;
++
++import io.papermc.paper.plugin.provider.type.PluginFileType;
++import org.slf4j.Logger;
++import org.slf4j.LoggerFactory;
++
++import java.io.IOException;
++import java.nio.file.Files;
++import java.nio.file.Path;
++import java.util.jar.JarFile;
++
++/*
++Responsible for storing all the plugin files.
++ */
++public class PluginProviders {
++
++    public static final PluginProviders INSTANCE = new PluginProviders();
++    private static final Logger LOGGER = LoggerFactory.getLogger(PluginProviders.class);
++
++    public void handlePluginDirectory(Path toPath) throws IOException {
++        if (!Files.isDirectory(toPath)) { // Create if missing
++            Files.createDirectories(toPath);
++        }
++
++        this.addDirectory(toPath);
++    }
++
++    public void addDirectory(Path directory) {
++        try {
++            Files.walk(directory, 1).filter(Files::isRegularFile).forEach(this::addFile);
++        } catch (Exception e) {
++            throw new RuntimeException(e);
```

Comment:
Just let this function throw it coz handlePluginDirection can already throw a ioexception

Reply:
This part has been redone